src/code/z_demo.c

```diff
@@ -30,27 +30,48 @@
 
 #include "assets/scenes/misc/hakaana_ouke/hakaana_ouke_scene.h"
 
-u16 D_8011E1C0 = 0;
-u16 D_8011E1C4 = 0;
-
-typedef void (*CutsceneStateHandler)(PlayState*, CutsceneContext*);
-
-void func_80064720(PlayState* play, CutsceneContext* csCtx);
-void func_80064760(PlayState* play, CutsceneContext* csCtx);
-void func_800647C0(PlayState* play, CutsceneContext* csCtx);
-void func_80068C3C(PlayState* play, CutsceneContext* csCtx);
-void func_80068D84(PlayState* play, CutsceneContext* csCtx);
-void func_80068DC0(PlayState* play, CutsceneContext* csCtx);
-
-CutsceneStateHandler sCsStateHandlers1[] = {
-    func_80064720, func_80064760, func_80064720, func_80068D84, func_80064720,
+u16 sCurTextId = 0;
+u16 sCurOcarinaAction = 0;
+
+typedef void (*CutsceneHandler)(PlayState*, CutsceneContext*);
+
+void CutsceneHandler_DoNothing(PlayState* play, CutsceneContext* csCtx);
+void CutsceneHandler_StartManual(PlayState* play, CutsceneContext* csCtx);
+void CutsceneHandler_StartScript(PlayState* play, CutsceneContext* csCtx);
+void CutsceneHandler_RunScript(PlayState* play, CutsceneContext* csCtx);
+void CutsceneHandler_StopManual(PlayState* play, CutsceneContext* csCtx);
+void CutsceneHandler_StopScript(PlayState* play, CutsceneContext* csCtx);
+
+CutsceneHandler sManualCutsceneHandlers[] = {
+    CutsceneHandler_DoNothing,   // CS_STATE_IDLE
+    CutsceneHandler_StartManual, // CS_STATE_START
+    CutsceneHandler_DoNothing,   // CS_STATE_RUN
+    CutsceneHandler_StopManual,  // CS_STATE_STOP
+    CutsceneHandler_DoNothing,   // CS_STATE_RUN_UNSKIPPABLE
 };
 
-CutsceneStateHandler sCsStateHandlers2[] = {
-    func_80064720, func_800647C0, func_80068C3C, func_80068DC0, func_80068C3C,
+CutsceneHandler sScriptedCutsceneHandlers[] = {
+    CutsceneHandler_DoNothing,   // CS_STATE_IDLE
+    CutsceneHandler_StartScript, // CS_STATE_START
+    CutsceneHandler_RunScript,   // CS_STATE_RUN
+    CutsceneHandler_StopScript,  // CS_STATE_STOP
+    CutsceneHandler_RunScript,   // CS_STATE_RUN_UNSKIPPABLE
 };
 
-u8 sTitleCsState = 0;
+typedef enum {
+    /* 0 */ TITLE_DEMO_SPIRIT_TEMPLE,
+    /* 1 */ TITLE_DEMO_DEATH_MOUNTAIN_CRATER,
+    /* 2 */ TITLE_DEMO_GANONDORF_HORSE
+} TitleDemoDestination;
+
+u8 sTitleDemoDestination = TITLE_DEMO_SPIRIT_TEMPLE;
+
+typedef struct {
+    /* 0x00 */ u16 entrance;      // entrance index upon which the cutscene should trigger
+    /* 0x02 */ u8 ageRestriction; // 0 for adult only, 1 for child only, 2 for both ages
+    /* 0x03 */ u8 flag;           // eventChkInf flag bound to the entrance cutscene
+    /* 0x04 */ void* script;      // pointer to the cutscene script
```

Comment:
```suggestion
    /* 0x04 */ void* script;      // segmented address pointing to the cutscene script
```

Reply:
it doesnt need to be segmented, plenty of actors keep cs data in their overlay

Reply:
this is in `EntranceCutscene`

a RAM address works as a segmented address too (segmented_to_virtual(ram) == ram)

what I mean matters here is that these go through segmented_to_virtual, so segmented addresses are valid

Reply:
yea ram address works fine in segmented_to_virtual, but I think its better to be more general about it to capture both cases

they only use segmented address for entrance cutscenes but it doesnt have to be. just saying pointer to the script is perfectly fine imo, it covers both cases of segmented and not

Reply:
> pointer covers both cases of segmented and not

No it doesn't

If it did we would have segmented_to_virtual wrapping every pointer usage

Reply:
a segmented address is still a pointer. a pointer is fine to describe it. it means nothing for the use of the segmented_to_virtual macro lol

Reply:
How do you refer to a segmented address with a segment number of 0 ?

Reply:
i dont understand what youre getting at
if you dont agree that a segmented address is a pointer, then you disagree with literally everywhere in the codebase where we type them as one

Reply:
"be a pointer" and "be typed as a pointer" is different to me yeah

Reply:
we'll need more input then cause i do not see it the same way you do

Reply:
To be clear I agree with the following:

Pointers are basically memory addresses (typically kseg0 0x80xx_xxxx)
We type stuff holding segmented addresses as pointers
I don't count segmented addresses as memory addresses (as in, part of the N64 memory map)
Segmented addresses need to be translated to memory addresses to be used as the typical C pointers
This warrants documenting what holds segmented addresses, since `data->stuff` doesn't work if `data` is a segmented address

Reply:
the field in question will still work even if the thing is not segmented. i think it is important to encapsulate both (without relying on the fact that seg_to_virt will return a virtual address back).
to me, segmented addresses are still pointers, and the term covers both well

if the field in question *had* to be a segmented address, then sure, I would agree to specifying. but does not have to be

Reply:
But you can't *have to* be a segmented address, that's explicitly a design choice of segmented addresses, "normal" memory addresses can be translated through segment address translating just fine, resulting in the same address

See https://github.com/Dragorn421/z64-romhack-tutorials/blob/master/segmented_addresses.md#segment-0
and the sources at the bottom of that document

Reply:
Example

```c
void Set(s32* redData, s32* greenData) {
    red = *redData;
    green = *greenData;
}

struct {
    s32* redData;
    s32* greenData;
} context;

Set(SEGMENTED_TO_VIRTUAL(context.redData), context.greenData);
```

redData can be a segmented address
greenData cannot
->
```c
struct {
    s32* redData; // segmented address
    s32* greenData;
} context;
```

Reply:
here, I removed a mention of pointers at all. sweep the problem under the rug and we'll argue about this exact same thing in another context some months from now 😂